File: math_engine/data_puller.py
import os
import sys
from pathlib import Path

from dotenv import load_dotenv
from supabase import Client, create_client

# math_engine/ sits directly under the repo root, so .env is one level up.
root_dir = Path(__file__).resolve().parent.parent
load_dotenv(dotenv_path=root_dir / ".env")


def get_supabase() -> Client:
    url = os.environ.get("SUPABASE_URL") or os.environ.get("NEXT_PUBLIC_SUPABASE_URL")
    key = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")
    if not url or not key:
        print("ERROR: Set SUPABASE_URL (or NEXT_PUBLIC_SUPABASE_URL) and "
              "SUPABASE_SERVICE_ROLE_KEY in .env.", file=sys.stderr)
        sys.exit(1)
    return create_client(url, key)


def fetch_investor_data() -> list[dict]:
    """
    Returns one dict per contact (individual investor) -- this is what actually gets
    scored and recommended now, not the firm. Each contact carries:
      - their personal deal history (contact_investments), with investment_stage /
        year_partnered borrowed from their firm's portfolio_investments for the same
        company, since contact_investments itself has neither column
      - their firm's aggregate vertical_focus / typical_check_size as a fallback,
        since individual partners don't have their own (contact_verticals exists but
        is currently unpopulated everywhere -- used when present, falls back to the
        firm's when not)
    """
    supabase = get_supabase()

    orgs_res = (
        supabase.table("organizations")
        .select(
            "id, aum, "
            "vertical_focus(preferred_stage, typical_check_size, verticals(vertical_name))"
        )
        .execute()
    )
    org_context = {o["id"]: o for o in orgs_res.data}

    pi_res = (
        supabase.table("portfolio_investments")
        .select("org_id, company_id, investment_stage, year_partnered")
        .execute()
    )
    stage_by_org_company = {(pi["org_id"], pi["company_id"]): pi for pi in pi_res.data}

    contacts_res = (
        supabase.table("contacts")
        .select(
            "id, first_name, last_name, role, linkedin_url, bio, org_id, "
            "organizations(name), "
            "contact_verticals(verticals(vertical_name)), "
            "contact_investments(relationship, company_id, companies(name, description))"
        )
        .execute()
    )

    contacts = []
    for c in contacts_res.data:
        org_id = c["org_id"]
        org = c.get("organizations") or {}

        investments = []
        for ci in c.get("contact_investments") or []:
            company = ci.get("companies") or {}
            borrowed = stage_by_org_company.get((org_id, ci.get("company_id")), {})
            investments.append({
                "company_name": company.get("name"),
                "description": company.get("description"),
                "relationship": ci.get("relationship"),
                "investment_stage": borrowed.get("investment_stage"),
                "year_partnered": borrowed.get("year_partnered"),
            })

        contacts.append({
            "id": c["id"],
            "name": f"{c['first_name']} {c['last_name']}".strip(),
            "role": c.get("role"),
            "linkedin_url": c.get("linkedin_url"),
            "bio": c.get("bio") or "",
            "org_id": org_id,
            "org_name": org.get("name"),
            "contact_verticals": [
                cv["verticals"]["vertical_name"]
                for cv in (c.get("contact_verticals") or [])
                if cv.get("verticals")
            ],
            "org_vertical_focus": (org_context.get(org_id) or {}).get("vertical_focus") or [],
            "investments": investments,
        })

    return contacts


def fetch_verticals() -> list[dict]:
    """All known verticals, for mapping a startup's free-text sector tags onto our taxonomy."""
    supabase = get_supabase()
    return supabase.table("verticals").select("id, vertical_name").execute().data


if __name__ == "__main__":
    data = fetch_investor_data()
    print(f"Fetched {len(data)} contacts.")
    if data:
        print(data[0])

File: lib/matching/dataPuller.ts
// Ported from math_engine/data_puller.py's fetch_investor_data().
import type { SupabaseClient } from "@supabase/supabase-js";
import type { Contact, VerticalFocusRow } from "./types";

function one<T>(v: T | T[] | null | undefined): T | null {
  if (!v) return null;
  return Array.isArray(v) ? (v[0] ?? null) : v;
}

// Returns one entry per contact (individual investor) -- this is what actually
// gets scored and recommended, not the firm. Each contact carries:
//   - their personal deal history (contact_investments), with investment_stage /
//     year_partnered borrowed from their firm's portfolio_investments for the same
//     company, since contact_investments itself has neither column
//   - their firm's aggregate vertical_focus / typical_check_size as a fallback,
//     since individual partners don't have their own (contact_verticals is used
//     when present, falls back to the firm's when not)
export async function fetchInvestorData(sb: SupabaseClient): Promise<Contact[]> {
  const orgsRes = await sb
    .from("organizations")
    .select("id, aum, vertical_focus(preferred_stage, typical_check_size, verticals(vertical_name))");
  if (orgsRes.error) throw new Error(orgsRes.error.message);

  const orgVerticalFocus = new Map<string, VerticalFocusRow[]>();
  for (const o of (orgsRes.data ?? []) as any[]) {
    const rows = (o.vertical_focus ?? []).map((vf: any) => ({
      preferred_stage: vf.preferred_stage ?? null,
      typical_check_size: vf.typical_check_size ?? null,
      verticals: one(vf.verticals),
    }));
    orgVerticalFocus.set(o.id, rows);
  }

  const piRes = await sb
    .from("portfolio_investments")
    .select("org_id, company_id, investment_stage, year_partnered");
  if (piRes.error) throw new Error(piRes.error.message);

  const stageByOrgCompany = new Map<string, { investment_stage: string | null; year_partnered: number | null }>();
  for (const pi of piRes.data ?? []) {
    stageByOrgCompany.set(`${pi.org_id}:${pi.company_id}`, {
      investment_stage: pi.investment_stage,
      year_partnered: pi.year_partnered,
    });
  }

  const contactsRes = await sb
    .from("contacts")
    .select(
      "id, first_name, last_name, role, linkedin_url, bio, org_id, typical_check_size, " +
        "organizations(name), " +
        "contact_verticals(verticals(vertical_name)), " +
        "contact_investments(relationship, company_id, companies(name, description))",
    );
  if (contactsRes.error) throw new Error(contactsRes.error.message);

  const contacts: Contact[] = [];
  for (const c of (contactsRes.data ?? []) as any[]) {
    const orgId = c.org_id;
    const org = one<{ name: string }>(c.organizations);

    const investments = (c.contact_investments ?? []).map((ci: any) => {
      const company = one<{ name: string; description: string | null }>(ci.companies);
      const borrowed = stageByOrgCompany.get(`${orgId}:${ci.company_id}`);
      return {
        company_name: company?.name ?? null,
        description: company?.description ?? null,
        relationship: ci.relationship ?? null,
        investment_stage: borrowed?.investment_stage ?? null,
        year_partnered: borrowed?.year_partnered ?? null,
      };
    });

    const contactVerticals = (c.contact_verticals ?? [])
      .map((cv: any) => one<{ vertical_name: string }>(cv.verticals)?.vertical_name)
      .filter((v: string | undefined): v is string => Boolean(v));

    contacts.push({
      id: c.id,
      name: `${c.first_name} ${c.last_name}`.trim(),
      role: c.role ?? null,
      linkedin_url: c.linkedin_url ?? null,
      bio: c.bio ?? "",
      org_id: orgId,
      org_name: org?.name ?? null,
      typical_check_size: c.typical_check_size ?? null,
      contact_verticals: contactVerticals,
      org_vertical_focus: orgVerticalFocus.get(orgId) ?? [],
      investments,
    });
  }

  return contacts;
}
